### PR TITLE
test: add the check for non-root user

### PR DESCRIPTION
**Pull request by yunmingyang**
- Requires https://github.com/cockpit-project/bots/pull/3297

### CODE REVIEW

File: test/check-machines-lifecycle [hunk 1]
```diff
@@ -38,9 +38,9 @@ class TestMachinesLifecycle(VirtualMachinesCase):
         self._testBasic()
 
     def createUser(self, user_group):
-        user_name = 'test_' + user_group + '_user'
-        self.machine.execute(
-            'useradd -G {0} {1} && echo "{1}:foobar" | chpasswd'.format(user_group, user_name))
+        user_name = 'test_{}_user'.format(user_group if user_group else "none")
```
Comment (martinpitt) [inline test/check-machines-lifecycle:41]: Please use f-strings, not `.format()`.

File: test/check-machines-lifecycle [hunk 2]
```diff
@@ -49,6 +49,10 @@ class TestMachinesLifecycle(VirtualMachinesCase):
 
         return user_name
 
+    def testBasicNonrootUserUnprivileged(self):
+        user = self.createUser(user_group=None)
+        self._testBasic(user, False, True)
```
Comment (martinpitt) [inline test/check-machines-lifecycle:54]: I don't quite understand how this can behave exactly the same way as testBasicWheelUserUnprivileged(). The latter ought to get superuser privileges, this one shouldn't. There must be some measurable difference. Actually, *this* new test case makes sense, I don't understand the existing testBasicWheelUserUnprivileged(). @yunmingyang , @skobyda , any thoughts here?

**Review by martinpitt — CHANGES_REQUESTED** [review]

File: test/check-machines-lifecycle (hunk 2 above)
Comment (skobyda) [reply test/check-machines-lifecycle:54]: So in order for wheel group user to have special privileges, you have to amend libvirt polkit policies. Without them, which is the case in our tests, wheel group user is basically any user without access to libvirt daemon. So I think testBasicNonrootUserUnprivileged is doing exactly the same as testBasicWheelUserUnprivileged.

What I think is missing in our tests instead is the extension for testBasicWheelUserUnprivileged which grants wheel group user priviliged access to qemu:///system by ammending polkit policies and see if it works properly. However there are currently som bugs with polkit, which need to be fixed first (e.g. https://bugzilla.redhat.com/show_bug.cgi?id=1820493), so that would be outside of the scope of this PR.

Comment (skobyda) [reply test/check-machines-lifecycle:54]: So in summary, I think testBasicNonrootUserUnprivileged doesn't add anything new compared to testBasicWheelUserUnprivileged. But I'm up for deleting testBasicWheelUserUnprivileged, since its name is kinda misleading, and reintroducing it back, with extended tests for privileged access for wheel group once the polkit issues are fixed.

**Review by skobyda — CHANGES_REQUESTED** [review]
I like this change, but let's also delete testBasicWheelUserUnprivileged since it doesn't do anything different and has a misleading name.

File: test/check-machines-lifecycle (hunk 2 above)
Comment (martinpitt) [reply test/check-machines-lifecycle:54]: What I meant is that a `wheel` user can run `sudo` and thus should be able to connect to libvirt as root. A test *can* disable this with `login_and_go(..., superuser=False)` option, but this test doesn't. That's where my confusion comes from -- why does this test not simply connect as root?

Comment (skobyda) [reply test/check-machines-lifecycle:54]: So I went back in git history and found that this seems to be rewritten from older test:

```
-    def testWheelUserOperationWithVm(self):
-        name = 'staticvm'
-        vm_args = self.create_vm(name, wait=True)
-
-        # A admin user in the wheel group is allowed to see the VM and
-        # can do operations on it, but only with escalated privileges.
-
-        self.non_root_user_operations(user_group='wheel', privilege=False,
-                                      operations=False)
-
-        self.non_root_user_operations(user_group='wheel', privilege=True,
-                                      operations=True, vm_args=vm_args)

```

Comment (skobyda) [reply test/check-machines-lifecycle:54]: So yeah I was wrong, you don't need polkit setup for that. But if you login with user with wheel privileges, you right now won't see any VMs (reproducible with `[wheeluser@fedora cockpit]$virsh -c qemu:///system list --all`) but once you authorize, you will get access to qemu:///system (reporducible with `[wheeluser@fedora cockpit]$sudo virsh -c qemu:///system list --all`). So I think the tests should reflect this behaviour

**Comment by skobyda** [comment]
@yunmingyang I added a [commit](https://github.com/cockpit-project/cockpit-machines/pull/669/commits/79b2fca15a3731f866fdd10e1db6ec9fbb0945ee) which extends wheel user test so it's tests more extensivily the expected behaviour for wheel user (as described in the [other comment](https://github.com/cockpit-project/cockpit-machines/pull/669#discussion_r853892904)):  What do you think about that?

File: test/check-machines-lifecycle [hunk 3]
```diff
@@ -60,9 +60,14 @@ class TestMachinesLifecycle(VirtualMachinesCase):
...
 
-    def testBasicWheelUserUnprivileged(self):
+    def testBasicWheelUser(self):
         user = self.createUser(user_group='wheel')
+
+        # Check NON-PRIVILEGED user won't see any VMs even though he is in wheel group
         self._testBasic(user, False, True)
+        self.machine.execute("virsh destroy subVmTest1 && virsh undefine subVmTest1")
+        # Check wheel group PRIVILEGED user will see VMs and can do operations on them
+        self._testBasic(user, True, False)
```
Comment (martinpitt) [inline test/check-machines-lifecycle:70]: That makes sense, thanks! Would be even nicer to name the boolean arguments, for easier interpretation. I.e. `superuser=True`

**Comment by yunmingyang** [comment]
@skobyda Yeah, sure, I also re-push the commit as modifying the f-string, please review, thanks.

File: test/check-machines-lifecycle (hunk 2 above)
Comment (yunmingyang) [reply test/check-machines-lifecycle:53]: I found we use self.eval_js('window.localStorage.setItem("superuser:%s", "%s");' % (user, "any" if superuser else "none")) to implement "superuser" in login_and_go. But it is "sudo" by window.localStorage.getItem("superuser:test_wheel") if changing the privilege in UI. Should modify "any" to "sudo"?

Comment (skobyda) [reply test/check-machines-lifecycle:53]: Ah I'm not really familiar with that code, or where and how it gets read by cockpit during authorization process. @martinpitt any idea?

Comment (martinpitt) [reply test/check-machines-lifecycle:53]: I'm afraid I don't understand @yunmingyang 's question. The `superuser` flag in `login_and_go()` determines whether the test will do the equivalent of clicking the "Limited access" button in the UI, so that the cockpit session will start a root bridge (through sudo) and hence all `{ superuser: "try" }` or `{ superuser: "require" }` cockpit API calls will run as root. Being in the `wheel` group allows a user to become root through sudo.

In other words, if you log in as `test_wheel_user` with `superuser=False`, this will behave like a normal unprivileged user. With `superuser=True` you will be able to do any administrative action including libvirt (this is totally independent of polkit, it just connects to libvirtd as root).

Comment (yunmingyang) [reply test/check-machines-lifecycle:53]: I mean, in the test, it seems to use "window.localStorage.setItem("superuser:%s", "%s")" in the browser console, before click "Log in", to do the equivalent of clicking the "Limited access" button in the UI. And if we use a wheel group user like test_wheel_user, to get "Administrative access", in test, it seems to use "window.localStorage.setItem("superuser:test_wheel_user", "any")". But if we get "Administrative access" by clicking ""Limited access", the value of "superuser:test_wheel_user" getting by window.localStorage.getItem("superuser:test_wheel_user") is "sudo", not "any". So my question is, is there some different when setting "superuser:test_wheel_user" to "sudo" or "any"?

Comment (martinpitt) [reply test/check-machines-lifecycle:53]: @yunmingyang : not in a meaningful way in this context, no. There used to be a fallback for running `pkexec cockpit-bridge` instead of `sudo cockpit-bridge`, but in practice, `any` and `sudo` are equivalent for "connect with browser" mode.

**Comment by yunmingyang** [comment]
@martin, @skobyda I check the failed job(debian-stable). It seems that /etc/sudoers in debian-stable lacks "%wheel  ALL=(ALL)       ALL", after adding that, the case could run continually. And I guess it is the same with debian-testing as their failed snapshots are the same. Should we update these two image by image-refresh? or I think we could also add a if in the test code.

Also, after the case run finished, I found there is a core dump by libvirtd and qemu

**Comment by yunmingyang** [comment]
According to: https://github.com/cockpit-project/bots/issues/3275, change testBasicWheelUser to testBasicAdminUser, and use "sudo" group for debian based distro

**Comment by martinpitt** [comment]
Thanks @yunmingyang ! I squashed the two topmost commits to get individually working commits in this PR series, and did a few grammar tweaks. I also rebased to current main so that the packit tests can run.

**Review by martinpitt — DISMISSED** [review]
LGTM once green.

**Comment by martinpitt** [comment]
Test failures on [debian-stable](https://logs.cockpit-project.org/logs/pull-669-20220425-101042-7a78907f-debian-stable/log.html) and [rhel-8-6](https://logs.cockpit-project.org/logs/pull-669-20220425-101054-7a78907f-rhel-8-6/log.html)

**Comment by martinpitt** [comment]
This needs to adjust the Testing Farm exclusions as well, [c9s already timed out](https://artifacts.dev.testing-farm.io/87d78f47-07cf-44cb-9883-c2bb87f92a37/).

I can't reproduce the rhel-8-6 failure, that just looks like a timeout on our busy CI (we run on a fallback today). 

The debian-stable crashes are real, and have long been known as https://github.com/cockpit-project/bots/issues/2220 . The naughty patterns need to be adjusted to also match the renamed and new test case.

**Comment by martinpitt** [comment]
I adjusted the pattern in https://github.com/cockpit-project/bots/pull/3297 , running test against that; also retrying rhel-8-6.

Once the bots PR is in, I'll rebase and adjust the TF tests, too.

**Comment by martinpitt** [comment]
The [testDelete](https://logs.cockpit-project.org/logs/pull-669-20220425-111653-7a78907f-debian-stable-bots-3297/log.html#73) failure happened again, seems the different timing on our AWS instances exposes this quite a bit. But I can reproduce this locally as well, I'll have a quick look.

**Review by martinpitt — APPROVED** [review]
LGTM now. @skobyda , can you please re-review as well?
